### PR TITLE
[MRG] Nested backend context manager

### PR DESCRIPTION
**Pull request by ogrisel**
While reviewing #516 (loky integration) and #537 (backend negotiation based on code hints / semantic constraints), I realized that the `with parallel_backend(backend_name):` idiom should support nesting when possible.

Here is a (currently failing) test to demonstrate what I mean. I will work on an implementation. If we ensure that backend instances should be picklable, that should be quite easy to implement although I am worried about the performance overhead it could induce.

### CODE REVIEW

File: joblib/test/test_parallel.py [hunk 1]
```diff
@@ -648,6 +645,50 @@ def test_direct_parameterized_backend_context_manager():
...
+
+    def nested_call(n):
+        return Parallel(n_jobs=2)(delayed(id)(i) for i in range(n))
+
+    with parallel_backend("retrieval") as (ba, _):
+        Parallel(n_jobs=2)(
+            delayed(nested_call, check_pickle=False)(i)
+            for i in range(5)
+        )
+        assert ba.i == 1
```
Comment (TomAugspurger) [inline joblib/test/test_parallel.py]: Does this look like the right number of calls to `retrieve`, given my `Parallel` calls?

File: joblib/_parallel_backends.py [hunk 2]
```diff
@@ -100,6 +101,30 @@ def abort_everything(self, ensure_ready=True):
...
+        Calls to Parallel.retrieve will be made inside this context.
+
+        By default, this does nothing. It may be useful for subclasses to
+        handle nested parallelism. In particular, it may be required to avoid
+        deadlocks if a backend manages a fixed number of workers, when those
+        workers may be asked to do nested Parallel calls. Without
+        'retrieval_context' this could lead to deadlock, as all the workers
+        managed by the backend may be "busy" waiting for the nested parallel
+        calls to finish, but the backend has no free workers to execute those
+        tasks.
```
Comment (ogrisel) [inline joblib/_parallel_backends.py]: Very nice explanation :)

File: joblib/test/test_parallel.py (hunk 1 above)
Comment (ogrisel) [reply joblib/test/test_parallel.py]: Yes, there is one only call to `retrieve` per call to `Parallel.__call__`.

File: joblib/test/test_parallel.py [hunk 3]
```diff
@@ -648,6 +645,50 @@ def test_direct_parameterized_backend_context_manager():
...
+            yield
+
+    register_parallel_backend("retrieval", MyBackend)
+
+    def nested_call(n):
+        return Parallel(n_jobs=2)(delayed(id)(i) for i in range(n))
+
+    with parallel_backend("retrieval") as (ba, _):
+        Parallel(n_jobs=2)(
+            delayed(nested_call, check_pickle=False)(i)
```
Comment (ogrisel) [inline joblib/test/test_parallel.py]: We can probably remove that `check_pickle=False` now that loky is the default backend and that cloudpickle is used to pickle function definitions.

**Comment by ogrisel** [comment]
Merged! Thanks @TomAugspurger.

**Pull request merged**